library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,683 @@
+
+    /// Records that the patch with number patch_number failed to boot, and ensures
+    /// that it will never be returned as the next boot or last booted patch.
+    fn record_boot_failure_for_patch(&mut self, patch_number: usize) -> Result<()>;
+
+    /// The highest patch number that has been added. This may be higher than the
+    /// last booted or next boot patch if we downloaded a patch that failed to boot.
+    fn highest_seen_patch_number(&self) -> Option<usize>;
+
+    /// Resets the patch manager to its initial state, removing all patches. This is
+    /// intended to be used when a new release version is installed.
+    fn reset(&mut self) -> Result<()>;
+}
+
+// This allows us to use the Debug trait on dyn ManagePatches, which is
+// required to have it as a property of UpdaterState.
+impl Debug for dyn ManagePatches {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "ManagePatches")
+    }
+}
+
+#[derive(Debug)]
+pub struct PatchManager {
+    /// The base directory used to store patch artifacts and state.
+    /// The directory structure created within this directory is:
+    ///  patches_state.json
+    ///  patches/
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    root_dir: PathBuf,
+
+    /// Metadata about the patches we have downloaded that is persisted to disk.
+    patches_state: PatchesState,
+}
+
+impl PatchManager {
+    /// Creates a new PatchManager with the given root directory. This directory is
+    /// assumed to exist. The PatchManager will use this directory to store its
+    /// state and patch binaries.
+    pub fn with_root_dir(root_dir: PathBuf) -> Self {
+        let patches_state = Self::load_patches_state(&root_dir).unwrap_or_default();
+
+        Self {
+            root_dir,
+            patches_state,
+        }
+    }
+
+    fn load_patches_state(root_dir: &Path) -> Option<PatchesState> {
+        let path = root_dir.join(PATCHES_STATE_FILE_NAME);
+        disk_io::read(&path).ok()
+    }
+
+    fn save_patches_state(&self) -> Result<()> {
+        let path = self.root_dir.join(PATCHES_STATE_FILE_NAME);
+        disk_io::write(&self.patches_state, &path)
+    }
+
+    /// The directory where all patch artifacts are stored.
+    fn patch_artifacts_dir(&self) -> PathBuf {
+        self.root_dir.join(PATCHES_DIR_NAME)
+    }
+
+    /// The directory where artifacts for the patch with the given number are stored.
+    fn dir_for_patch_number(&self, patch_number: usize) -> PathBuf {
+        self.patch_artifacts_dir().join(patch_number.to_string())
+    }
+
+    /// The path to the runnable patch artifact with the given number. Runnable patch artifact files are
+    /// named <patch_number>.vmcode
+    fn patch_artifact_path(&self, patch_number: usize) -> PathBuf {
+        self.dir_for_patch_number(patch_number).join("dlc.vmcode")
```

Comment:
Should dlc.vmcode be a constant somewhere?

Reply:
Done